### PR TITLE
[release-0.65] Add alert to notify of nmstate removal

### PR DESCRIPTION
**Pull request by rhrazdil**
kubernetes-nmstate is removed in the next CNAO version.
This change adds an alert that notifies users who have
knmstate deployed with CNAO, and point them to runbook,
that explains that standalone kubernetes-nmstate operator
should be installed.

Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>

**What this PR does / why we need it**:

**Special notes for your reviewer**:

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by RamLavi** [comment]
@sradco can you also review this ?

**Review by RamLavi — CHANGES_REQUESTED** [review]
Can you also add an e2e to alerts_tests?

File: test/e2e/workflow/deployment_test.go
Comment (rhrazdil) [reply test/e2e/workflow/deployment_test.go:411]: It also means adding the monitoring lane, right?

Comment (rhrazdil) [reply test/e2e/workflow/deployment_test.go:355]: Done

File: data/monitoring/prom-rule.yaml
Comment (yaacov) [inline data/monitoring/prom-rule.yaml:27]: we want to highlight this alert in the UI virtualization overview page.

can we add a label that will tell the UI to show this alert on the top of the virtualization overview page for better visibility ?

for exampe:
`kubevirt_ui: highlighted`
does that makes sense?

Comment (yaacov) [reply data/monitoring/prom-rule.yaml:27]: do we have a naming convention for alert labels?

Comment (rhrazdil) [reply data/monitoring/prom-rule.yaml:27]: I don't think we have, but @sradco may confirm that.

Comment (rhrazdil) [reply data/monitoring/prom-rule.yaml:27]: I would suggest

```
kubevirt_console_scope: [cluster,virtualmachine]
```
But IIUC, we need to provide more info in case an alert is cluster scoped (that's not case of the alert in this PR, just to make the mechanism general).

Comment (sradco) [reply data/monitoring/prom-rule.yaml:27]: @yaacov Can you please explain what is the reasoning to highlight this specific alert? 
I believe alerts should be ordered based on their severity.

Comment (yaacov) [reply data/monitoring/prom-rule.yaml:27]: `kubevirt_console_scope: [cluster,virtualmachine]`
is useful to specify the scope,
we will also need a way to specify that we want this alert to be shown in the top of the overview page.
`kubevirt_console_popup: [cluster,virtualmachine]`
?

Comment (rhrazdil) [reply data/monitoring/prom-rule.yaml:27]: @sradco We want to show this specific alert in OCP Console UI as a banner on CNV overview page

Comment (rhrazdil) [reply data/monitoring/prom-rule.yaml:27]: @yaacov My thinking was that is the kubevirt_console_scope is define, then we can implicitly assume that we want to show that alert in UI, no? Otherwise, if it's not there, UI wouldn't show it

Comment (sradco) [reply data/monitoring/prom-rule.yaml:27]: @yaacov please see my question above. I don;t understand the need for this.
Having these "console" details in the alert as labels will be wrong practice. 
If you want to do this, it should be in the UI level only.
The alerts can be used is multiple envs, not only OCP UI and are also sent to the user with alert manager with all its labels. 
Adding this will create clutter that is not relevant to the alert itself.
The new labels we added about the operator name and component are directly related to the alert details and help the users to understand where the issue comes from.

Comment (yaacov) [reply data/monitoring/prom-rule.yaml:27]: @sradco hi,

> the reasoning to highlight this specific alert?

IIUC the request is that an alert be shown on OCP Console UI as a banner on CNV overview page.
UI need some hint that this alert request to be shown this way.
( all alerts including this one will be shown in the regular alerts list as well. )

> If you want to do this, it should be in the UI level only.

AFAIU this is a property of the alert, the UI read this hint, and decide to show alerts that has this property differently.
this alert can then be updated by the people who create the alert rule, and decide to not show it if not needed anymore, or create a different alert that request to be displayed on top of the page.

It's an independent property from the severity, for example an alert can by warning or information and request to be highlighted.

Comment (yaacov) [reply data/monitoring/prom-rule.yaml:27]: @sradco
do we currently have a way to know that an alert belong to a specific virtual machine?

AFAIU the request was that this alert will be shown in the virtualization overview, and another for a specific vm that needs some action because of some event.

Comment (sradco) [reply data/monitoring/prom-rule.yaml:27]: I still think this should be done on the UI level. 
Perhaps there could be a place to state the alerts to show in the banner that is separate from the alerts . Then the UI can check based on it.
I strongly believe we should not  add these labels to the alerts.

Comment (sradco) [reply data/monitoring/prom-rule.yaml:27]: We don't have alerts for specific VMs at this point. There are planned alerts for 4.12.

Comment (ronensdeor) [reply data/monitoring/prom-rule.yaml:27]: @sradco We need a way to surface specific alerts so they are not "lost" in all the other alerts.
Those alert may have different severity, like a reboot is required, which can be warning, but the user must be aware of it.
The the current case, users that are using NMstate from CNV must install the stand alone operator before moving to 4.11. In this case it's not a critical alert the requires immediate action, but it is an important one. Having it as a warning will probably not be visible enough.

Comment (rhrazdil) [reply data/monitoring/prom-rule.yaml:27]: @sradco We're looking for a generic way of displaying an alert in UI (not in Observe/metrics).
Marking an alert with an annoration/label/something allows leaving all logic for the backend, and UI only presents
what it's supposed to.
Wiring logic into UI, what is supposed to be displayed, what not, is very sub-optimal.

> I strongly believe we should not add these labels to the alerts.
Could you explain please?

I believe having an extra annotation would be a huge clutter for the users.
At the very least, the gain we get - a user being notified on a relevant place in the UI, instead of somewhere in Monitoring category, seems to be prevalent to me

Comment (yaacov) [reply data/monitoring/prom-rule.yaml:27]: > Perhaps there could be a place to state the alerts to show in the banner that is separate from the alerts.

Also works for me, my concern is that maintainers that will update the alert rules forget to update the alert rules "impact to user" table that will be in a different file.
It will also be easier to a user looking at the active alert to filter by "impact to user", in case they don't use the UI.

Comment (sradco) [reply data/monitoring/prom-rule.yaml:27]: In that case I think we should not use UI in the label name.
It should be somehow reflected that an action is required or something like that.
Like a sub label to severity.

Comment (yaacov) [reply data/monitoring/prom-rule.yaml:27]: > It should be somehow reflected that an action is required

`admin_action_recommended: <some longer description>` 
or 
`recommended_action: yes/[no]` 

does that makes more sense?

Comment (sradco) [reply data/monitoring/prom-rule.yaml:27]: I thought of something like:
severity_score / severity_level / severity_priority with  a few optional values, but not sure what they should be called.

Comment (yaacov) [reply data/monitoring/prom-rule.yaml:27]: > Like a sub label to severity.

`severity_priority` makes sense to me because it indicate that handling this alert requires higher priority even if it's currently only a warning.

> What will happen if there are multiple ones?

They will all show in the top of the page, if they all require user to be aware of them, without raising them to critical and without hiding theme between all the other warnings and information alerts.

Comment (yaacov) [reply data/monitoring/prom-rule.yaml:27]: maybe event just `priority` , without the `sevirity_`  prefix ?

File: data/monitoring/prom-rule.yaml
Comment (rhrazdil) [reply data/monitoring/prom-rule.yaml:30]: We want to highlight the CnaoNmstateMigration alert, but yes, I get your point

File: data/monitoring/prom-rule.yaml
Comment (yaacov) [reply data/monitoring/prom-rule.yaml:27]: we will also need to document how to use the labels, we don't want every critical alert to also be high priority (because it's critical)

Comment (sradco) [reply data/monitoring/prom-rule.yaml:27]: I found a nice article that supports the naming and values for the label:
https://www.guru99.com/defect-severity-in-software-testing.html

I think we should keep it as priority.

Comment (sradco) [reply data/monitoring/prom-rule.yaml:27]: The documentation for the label should be part of OCP metrics documentation.
Im not sure if we should have it as mandatory as the severity. 
The issue is that the label naming impacts also OCP and all other operators.
We should propose this in the monitoring forum and get their opinion.

Comment (yaacov) [reply data/monitoring/prom-rule.yaml:27]: > I think we should keep it as priority.

makes sense to me too

> Im not sure if we should have it as mandatory as the severity.

IMHO it should not be mandatory (default to medium if missing) 
my reasoning is that unlike severity that is clear to define by the rule maintainer, priority may depend on things outside the scope of the alert rules maintainer, the cases where the rule maintainer wish to raise/lower priority are the odd cases.

Comment (sradco) [reply data/monitoring/prom-rule.yaml:27]: I created https://github.com/openshift/enhancements/pull/1077. Please review.

Comment (yaacov) [reply data/monitoring/prom-rule.yaml:27]: from https://github.com/openshift/enhancements/pull/1077 is looks like "priority: high"

@rhrazdil can you add the "priority" label?

**Comment by phoracek** [comment]
Adding explicit hold so we don't merge this preemptively

/hold

File: data/monitoring/prom-rule.yaml
Comment (rhrazdil) [reply data/monitoring/prom-rule.yaml:28]: Sorry, I've missed your comment.
Added the `priority: high` label in a new commit

File: data/monitoring/prom-rule.yaml
Comment (sradco) [inline data/monitoring/prom-rule.yaml:25]: @rhrazdil Please move the Priority label below the severity.

Comment (yaacov) [reply data/monitoring/prom-rule.yaml:25]: IMHO semantically, the required fields should go before the optional ones,
using that way of thinking, `severity` that is  required label (and more important) will go before the optional `priority`.
I have nothing against going with alphabetical order if it's the common practice for labels, so that is also good option, don't know what is best.

Comment (sradco) [reply data/monitoring/prom-rule.yaml:25]: @rhrazdil I agree with Kobi.

Comment (rhrazdil) [reply data/monitoring/prom-rule.yaml]: Okay, I take it. [Updated](https://github.com/kubevirt/cluster-network-addons-operator/compare/46cadb76d39bfc90514006a71397e8553100edef..faf97e0175d2f1fa3bd071b60a3f44fcccd668ca)

File: pkg/monitoring/metrics.go
Comment (RamLavi) [inline pkg/monitoring/metrics.go:38]: hmmm.. I don't get something..
if you already defined `kubevirt_cnao_nmstate_deployed` with `kube_daemonset_labels`, i.e. without the use of a GAUGE variable, why do you need to define one here? isn't this redundant?

**Review by RamLavi — CHANGES_REQUESTED** [review]
Please check if we really need to define `kubevirt_cnao_nmstate_deployed` as gauge, if you can define it with pure prometheus given metrics (like it seems you already have succeeded), then we can remove the gauges, which will allow us to remove some of the cherry picks which will make things simpler IMO)

File: pkg/monitoring/metrics.go
Comment (rhrazdil) [reply pkg/monitoring/metrics.go]: Removed the variable in this [force-push](https://github.com/kubevirt/cluster-network-addons-operator/compare/faf97e0175d2f1fa3bd071b60a3f44fcccd668ca..7cc8bcd687897940a33fddda5cb0a38b1f71031a).

**Comment by rhrazdil** [comment]
> Please check if we really need to define kubevirt_cnao_nmstate_deployed as gauge, if you can define it with pure prometheus given metrics (like it seems you already have succeeded), then we can remove the gauges, which will allow us to remove some of the cherry picks which will make things simpler IMO)

Updated (removed the redundant variable).
To the cherry-picks, the monitoring lane adds infra for testing alerts, we could add the infra alone without adding the lane, but it seems easier from maintenance perspective to cherry-pick the whole commit.
The following two commits:
[monitoring, e2e, Move endpoints test to monitoring lane ](https://github.com/kubevirt/cluster-network-addons-operator/pull/1302/commits/2ce50185e827c006c24403bc9e60246f5fb7e37b) 
[tests, metrics: Fix potential flakes on metrics scraping test](https://github.com/kubevirt/cluster-network-addons-operator/pull/1302/commits/b0cd67ef4ca55999e4585f96325c9bd15f8faf45)

Are not really neccessary, we could keep the tests in workflow/deployment, if you prefer

**Comment by yaacov** [comment]
@rhrazdil hi,
should we merge the alert in the main branch too ?
it makes sense to me to have fixes in main when possible ? 

@RamLavi WDYT ?

cc:// @sradco @phoracek

**Comment by rhrazdil** [comment]
Hello @yaacov,
we decided to only throw the alert in the release-0.65 branch, because cnao in newer release deletes kubernetes-nmstate, if it's installed.
So the alert couldn't fire in newer branch anyway.

**Comment by sradco** [comment]
@rhrazdil I think it is important to add this alerts also to main branch. In case the user upgrade directly to the next majot version and not to the next minor z-stream.

**Comment by rhrazdil** [comment]
We're blocking an upgrade when user upgrades from CNV 4.10 to CNV 4.11 and has kubernetes-nmstate deployed
via CNAO.
Hence, if user has CNV 4.11 (and thus also ocp 4.11) and knmstate deployed, we can safely assume, that it's deployed via Kubernetes Nmstate Operator. CNAO would delete the knmstate deployment within a minute.

So this alert has no meaning on 4.11

**Pull request merged**
